internal/common/schemafunc/find_changes.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type AttributeChanges struct {
+	Changes []string
+}
+
+func (a *AttributeChanges) LeafChanges() map[string]bool {
+	return a.leafChanges(true)
+}
+
+func (a *AttributeChanges) AttributeChanged(name string) bool {
+	changes := a.LeafChanges()
+	changed := changes[name]
+	return changed
+}
+
+func (a *AttributeChanges) KeepUnknown(attributeEffectedMapping map[string][]string) []string {
+	keepUnknown := []string{}
+	for attrChanged, affectedAttributes := range attributeEffectedMapping {
+		if a.AttributeChanged(attrChanged) {
+			keepUnknown = append(keepUnknown, attrChanged)
+			keepUnknown = append(keepUnknown, affectedAttributes...)
+		}
+	}
+	return keepUnknown
+}
+
+// ListIndexChanged returns true if the list at the given index has changed, false if it was added or removed
+func (a *AttributeChanges) ListIndexChanged(name string, index int) bool {
+	leafChanges := a.leafChanges(false)
+	indexPath := fmt.Sprintf("%s[%d]", name, index)
+	return leafChanges[indexPath]
+}
+
+// NestedListLenChanges accepts a fullPath, e.g., "replication_specs[0].region_configs" and returns true if the length of the nested list has changed
+func (a *AttributeChanges) NestedListLenChanges(fullPath string) bool {
+	addPrefix := fmt.Sprintf("%s[+", fullPath)
+	removePrefix := fmt.Sprintf("%s[-", fullPath)
+	for _, change := range a.Changes {
+		if strings.HasPrefix(change, addPrefix) || strings.HasPrefix(change, removePrefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *AttributeChanges) ListLenChanges(name string) bool {
+	leafChanges := a.leafChanges(false)
+	addPrefix := fmt.Sprintf("%s[+", name)
+	removePrefix := fmt.Sprintf("%s[-", name)
+	for change := range leafChanges {
+		if strings.HasPrefix(change, addPrefix) || strings.HasPrefix(change, removePrefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *AttributeChanges) leafChanges(removeIndex bool) map[string]bool {
+	leafChanges := map[string]bool{}
+	for _, change := range a.Changes {
+		var leaf string
+		parts := strings.Split(change, ".")
+		if len(parts) == 1 {
+			leaf = parts[0]
+		} else {
+			leaf = parts[len(parts)-1]
+		}
+		if removeIndex && strings.HasSuffix(leaf, "]") {
+			leaf = strings.Split(leaf, "[")[0]
+		}
+		leafChanges[leaf] = true
+	}
+	return leafChanges
+}
+
+// FindAttributeChanges: Iterates through TFModel of state+plan and returns AttributeChanges for querying changed attributes
+// The implementation is similar to KeepUnknown, no support for types.Set or types.Tuple yet
+func FindAttributeChanges(ctx context.Context, src, dest any) AttributeChanges {
+	changes := FindChanges(ctx, src, dest)
+	return AttributeChanges{changes}
+}
+
+func FindChanges(ctx context.Context, src, dest any) []string {
+	valSrc, valDest := validateStructPointers(src, dest)
+	typeDest := valDest.Type()
+	changes := []string{} // Always return an empty list, as nested attributes might be added and then removed
```

Comment:
```suggestion
	changes := nil // Always return an empty list, as nested attributes might be added and then removed
```

can we return nil? it's normally preferred to an explicit empty slice

Reply:
Updated the comment. Does it answer the question?

Reply:
do you mean this comment?:  , which make the test cases fail on nil vs []

I'm not sure yet, what's the difference between nil and empty string here?

Reply:
Yes, that is the one I mean. It was simpler to keep it always `string[]{}` to avoid the failing test assertion

Reply:
ah, ok, i get it, ok as it is